RealmSwift/ThreadSafeReference.swift

```diff
@@ -122,6 +122,72 @@ public protocol ThreadConfined {
     }
 }
 
+/**
+ Object types conforming to `ThreadSafeWrappable` are able to have the `@ThreadSafe` property wrapper applied to them.
+
+ Note that only types defined by Realm can meaningfully conform to this protocol, and defining new
+ classes which attempt to conform to it will not make them work with `@ThreadSafe`
+ */
+public protocol ThreadSafeWrappable: ThreadConfined {
+    /** The Realm property type associated with this type */
+    static var confinedType: PropertyType { get }
+}
+/**
+    A property type wrapper type that may be passed between threads.
+
+    A `@ThreadSafe` property contains a thread-safe reference to the underlying wrapped value. This reference is resolved to the thread on which the wrapped value is accessed. A new thread safe reference is created each time the property is accessed.
+
+ - see: `ThreadConfined`
+ - see: `ThreadSafeReference`
+*/
+@propertyWrapper public class ThreadSafe<T: ThreadSafeWrappable> {
+    var threadSafeReference: ThreadSafeReference<T>?
+    var configuration: Realm.Configuration?
+
+    /// :nodoc:
+    public var wrappedValue: T? {
+        get {
+            guard let threadSafeReference = threadSafeReference, let config = configuration else { return nil }
+            do {
+                let realm = try Realm(configuration: config)
+                guard let value = threadSafeReference.resolve(in: realm) else {
+                    self.threadSafeReference = nil
+                    return nil
+                }
+                self.threadSafeReference = ThreadSafeReference(to: value)
+                return value
+            // FIXME: wrappedValue should throw
+            // As of Swift 5.5 property wrappers can't have throwing accessors.
+            } catch let error as NSError { throwRealmException(error.localizedDescription) }
+        }
+        set {
+            guard let newValue = newValue else {
+                threadSafeReference = nil
+                return
+            }
+            guard let configuration = newValue.realm?.configuration else {
+                throwRealmException("No realm configuration on this object found. Only managed objects may be wrapped as thread safe.")
```

Comment:
The first half of this error message is just an implementation detail. The user doesn't know or care that we happen to grab the configuration off of the Realm, so there's no reason for the exception to mention that.